samcli/local/docker/container.py

```diff
@@ -378,11 +383,22 @@ def wait_for_http_response(self, name, event, stdout) -> Tuple[Union[str, bytes]
         # NOTE(sriram-mv): There is a connection timeout set on the http call to `aws-lambda-rie`, however there is not
         # a read time out for the response received from the server.
 
-        resp = requests.post(
-            self.URL.format(host=self._container_host, port=self.rapid_port_host, function_name="function"),
-            data=event.encode("utf-8"),
-            timeout=(self.RAPID_CONNECTION_TIMEOUT, None),
-        )
+        # generate a lock key with host-port combination which is unique per function
+        lock_key = f"{self._container_host}-{self.rapid_port_host}"
+        LOG.debug("Getting lock for the key %s", lock_key)
+        with CONCURRENT_CALL_MANAGER_LOCK:
```

Comment:
Why do we need this lock and not just the `CONCURRENT_CALL_MANAGER`?

Reply:
We need locks for each lambda container, and in order to create/access that lock we need another one. So `CONCURRENT_CALL_MANAGER_LOCK` locks the first call to either create/access the lock for certain lambda function. And that accessed/created lock will actually lock the RIE call below.